src/engine/GroupBy.cpp

```diff
@@ -677,6 +697,126 @@ bool GroupBy::computeOptimizedGroupByIfPossible(IdTable* result) {
+                                             const IdTable& subresult,
+                                             size_t columnIndex,
+                                             LocalVocab& localVocab) {
+  // TODO: Use an interface for AggregationData
+  ad_utility::HashMap<ValueId, AverageAggregationData> map;
+
+  // Initialize evaluation context
+  sparqlExpression::EvaluationContext evaluationContext(
+      *getExecutionContext(), _subtree->getVariableColumns(),
+      subresult,getExecutionContext()->getAllocator(), localVocab);
+
+  evaluationContext._groupedVariables = ad_utility::HashSet<Variable>{
+      _groupByVariables.begin(), _groupByVariables.end()};
+  // TODO: Save previous results (?)
+  evaluationContext._variableToColumnMapPreviousResults =
+      getInternallyVisibleVariableColumns();
+  evaluationContext._previousResultsFromSameGroup.resize(getResultWidth());
+  evaluationContext._isPartOfGroupBy = true;
+
+  size_t blockSize = 65536;
+
+  auto numBlocks = (size_t) ceil(((double) subresult.size()) / blockSize);
+
+  size_t overallResults = 0;
+
+  for (size_t i = 0; i < numBlocks; ++i) {
+    evaluationContext._beginIndex = i * blockSize;
+    evaluationContext._endIndex = std::min((i + 1) * blockSize, subresult.size());
+
+    auto currentBlockSize = evaluationContext._endIndex - evaluationContext._beginIndex;
+
+    std::vector<sparqlExpression::detail::IntOrDouble> resultValues;
+
+    // Is this slower than emplace_back?
+    // Calling IntOrDouble constructor twice VS. resizing of vector on the go
+    resultValues.resize(currentBlockSize);
+
+    // Evaluate child expression on block
+    auto expr = aggregates.at(0)._expression.getPimpl();
+    auto exprChildren = expr->children();
+
+    sparqlExpression::ExpressionResult expressionResult =
+        exprChildren[0]->evaluate(&evaluationContext);
+
+    auto visitor = [&]<sparqlExpression::SingleExpressionResult T>(
+                       T&& singleResult) mutable {
+      auto generator = sparqlExpression::detail::makeGenerator(
+          std::forward<T>(singleResult), currentBlockSize,
+          &evaluationContext);
+
+      using NVG = sparqlExpression::detail::NumericValueGetter;
+
+      size_t k = 0;
+      for (auto val : generator) {
+        sparqlExpression::detail::NumericValue numVal = NVG()(val, &evaluationContext);
+
+        if (const int64_t* pInt = std::get_if<int64_t>(&numVal))
+          resultValues[k] = *pInt;
+        else if (const double* pDouble = std::get_if<double>(&numVal))
+          resultValues[k] = *pDouble;
+        else
+          AD_FAIL();
+        ++k;
+      }
+      overallResults += k;
+    };
+
+    std::visit(visitor, std::move(expressionResult));
+
+    for (size_t j = 0; j < currentBlockSize; ++j) {
+      auto rowIndex = i * blockSize + j;
+      auto id = subresult(rowIndex, columnIndex);
+      map[id].increment(resultValues[j]);
+    }
```

Comment:
I think you can (and should) call the `increment` function directly inside the `for` loop over the generator without storing in a separate vector.

You can also store the current rowIndex, then you can simply increment it instead of computing it new for each iteration.